_posts/2021-04-14-accessing-the-relay-store.markdown

```diff
@@ -0,0 +1,260 @@
+</figure>
+
+The feature makes use of [Redis][about-redis], a key-value store used for in-memory cache, to store two lists:
+
+1. `includeIDs` to store the five artworks in the list, so users see a consistent list of artworks whenever they
+   log in and load the page
+2. `excludeIDs` or “snoozed” IDs which Redis will store for 24 hours and ensure the user does not see
+
+When a user presses the “snooze” button, the ID for the artwork is added to the snoozed list in Redis. The list
+`includeIDs` and the list of `excludeIDs` are passed down from Rails controllers to our HAML views and then passed
+as props into our React `HomePageChecklist` app. In our Checklist component, we use both the `includeIDs` and the
+`excludeIDs` as arguments passed to our Relay query to determine what is returned from Metaphysics (Artsy's GraphQL
+layer).
+
+```js
+fragment ArtworksMissingMetadata_partner on Partner
+@argumentDefinitions(
+  first: { type: "Int", defaultValue: 5 }
+  after: { type: "String" }
+  checklistArtworkIDs: { type: "[String!]" }
+  checklistExcludedArtworkIDs: { type: "[String!]" }
+  missingPriorityMetadata: { type: "Boolean", defaultValue: true }
+  publishedWithin: { type: "Int", defaultValue: 7776000 }
+  shallow: { type: "Boolean", defaultValue: false }
+) {
+  id
+  artworksConnection(
+    first: $first
+    after: $after
+    artworkIDs: $checklistArtworkIDs
+    exclude: $checklistExcludedArtworkIDs
+    missingPriorityMetadata: $missingPriorityMetadata
+    publishedWithin: $publishedWithin
+    shallow: $shallow
+  ) @connection(key: "ArtworksMissingMetadata_partner_artworksConnection", filters: []) {
+    totalCount
+    edges {
+      node {
+        internalID
+        ...ArtworksMissingMetadataItem_artwork
+      }
+    }
+  }
+}
+```
+
+## The Damn User
+
+The problem we were running into is that when the user presses “snooze” on an item, Redis is updated but the list
+that we’re using in our Relay query becomes stale. When the user reloads the page, the response from our Relay
+query is correct because it's using the fresh Redis list, but we needed to make sure that the list in the UI is
+also correct and does not include that snoozed item.
+
+Without reloading, our initial solution was to use a local state variable in our React component to keep track of
+which items were snoozed. This looked like:
+
+```js
+const [localSnoozedItems, setLocalSnoozedItems] = useState([])
+```
+
+`localSnoozedItems ` and `setLocalSnoozedItems` were passed down to each of the children items so when the “snooze”
+button was pressed on that item, not only was the new key stored in Redis (for future reloads of the page) but the
+`localSnoozedItems` state variable was also updated. The connection returned from our Relay query (which remember,
+is already filtered based on our Redis `excludeIDs` from Redis) passes through an additional check to make sure
+that none of those items were snoozed locally and checked again the `localSnoozedItemsList`.
```

Comment:
I'm struggling a bit with this paragraph. Each of the sentences has a few things going on -- is there any way you could make this point with more, smaller sentences?